File: leetcode/medium/838_push_dominoes.py
# -----------------------------------------
# My Solution
#
# Time  Complexity: O(n)
# Space Complexity: O(n)
# -----------------------------------------
# n := len(dominoes)
class Solution:
    def pushDominoes(self, dominoes: str) -> str:
        if len(dominoes) == 0:
            return ""

        to_right_force = ["."] * len(dominoes)
        cursor = 0
        while cursor < len(dominoes):
            if dominoes[cursor] == "R":
                while cursor < len(dominoes) and dominoes[cursor] != "L":
                    to_right_force[cursor] = "R"
                    cursor += 1
            else:
                cursor += 1

        to_left_force = ["."] * len(dominoes)
        cursor = len(dominoes) - 1
        while cursor >= 0:
            if dominoes[cursor] == "L":
                while cursor >= 0 and dominoes[cursor] != "R":
                    to_left_force[cursor] = "L"
                    cursor -= 1
            else:
                cursor -= 1

        output = ["."] * len(dominoes)
        cursor = 0
        while cursor < len(dominoes):
            if to_right_force[cursor] == "R" and to_left_force[cursor] == ".":
                output[cursor] = "R"
                cursor += 1
            elif to_right_force[cursor] == "." and to_left_force[cursor] == "L":
                output[cursor] = "L"
                cursor += 1
            elif to_right_force[cursor] == "R" and to_left_force[cursor] == "L":
                end_cursor = cursor
                while end_cursor + 1 < len(dominoes) and to_right_force[end_cursor + 1] == "R":
                    end_cursor += 1

                middle_cursor = (cursor + end_cursor) // 2
                # e.g.
                # "R..L"  -> "RRLL"
                # "R...L" -> "RR.LL"
                for index in range(cursor, middle_cursor + (1 if (end_cursor - cursor) % 2 == 1 else 0)):
                    output[index] = "R"
                for index in range(middle_cursor + 1, end_cursor + 1):
                    output[index] = "L"
                cursor = end_cursor + 1
            else:
                cursor += 1
        return "".join(output)

# -----------------------------------------
# Model Solution
#
# Time  Complexity: O(n)
# Space Complexity: O(n)
# -----------------------------------------
# n := len(dominoes)
# Ref: https://leetcode.com/problems/push-dominoes/solution/
class Solution:
    def pushDominoes(self, dominoes: str) -> str:
        n = len(dominoes)
        forces = [0] * n

        curr_force = 0
        for index in range(n):
            if dominoes[index] == "R":
                curr_force = n
            elif dominoes[index] == "L":
                curr_force = 0
            else:
                curr_force = max(curr_force - 1, 0)
            forces[index] += curr_force

        curr_force = 0
        for index in range(n - 1, -1, -1):
            if dominoes[index] == "R":
                curr_force = 0
            elif dominoes[index] == "L":
                curr_force = -n
            else:
                curr_force = min(curr_force + 1, 0)
            forces[index] += curr_force

        return "".join("." if force == 0 else ("R" if force > 0 else "L") for force in forces)

File: leetcode/medium/838_push_dominoes.cpp
// -----------------------------------------
// My Solution
//
// Time  Complexity: O(n)
// Space Complexity: O(n)
// -----------------------------------------
// n := dominoes.size()
class Solution {
public:
    string pushDominoes(string dominoes) {
        int max_force = dominoes.size();

        vector<int> positive_net_forces(dominoes.size());
        for (int index = 0; index < dominoes.size(); ++index) {
            if (dominoes[index] == 'R')
                positive_net_forces[index] = max_force;
            else if (dominoes[index] != 'L' && index > 0 && positive_net_forces[index - 1] > 0)
                positive_net_forces[index] = positive_net_forces[index - 1] - 1;
        }

        vector<int> negative_net_forces(dominoes.size());
        for (int index = dominoes.size() - 1; index >= 0; --index) {
            if (dominoes[index] == 'L')
                negative_net_forces[index] = -max_force;
            else if (dominoes[index] != 'R' && index < dominoes.size() - 1 && negative_net_forces[index + 1] < 0)
                negative_net_forces[index] = negative_net_forces[index + 1] + 1;
        }

        string final_state_string(dominoes.size(), '.');
        for (int index = 0; index < dominoes.size(); ++index) {
            int net_force = positive_net_forces[index] + negative_net_forces[index];
            if (net_force != 0)
                final_state_string[index] = net_force > 0 ? 'R' : 'L';
        }
        return final_state_string;
    }
};

// -----------------------------------------
// Model Solution: Two Pointers
//
// Time  Complexity: O(n)
// Space Complexity: O(1)
// -----------------------------------------
// n := dominoes.size()
//
// Ref: https://leetcode.com/problems/push-dominoes/discuss/132332/JavaC%2B%2BPython-Two-Pointers
class Solution {
public:
    string pushDominoes(string dominoes) {
        dominoes = 'L' + dominoes + 'R';

        string final_state;
        for (int i = 0, j = 1; j < dominoes.length(); ++j) {
            if (dominoes[j] == '.')
                continue;

            int distance = j - i - 1;
            if (i > 0)
                final_state += dominoes[i];
            if (dominoes[i] == dominoes[j])
                final_state += string(distance, dominoes[i]);
            else if (dominoes[i] == 'L' && dominoes[j] == 'R')
                final_state += string(distance, '.');
            else
                final_state += string(distance / 2, 'R') + string(distance % 2, '.') + string(distance / 2, 'L');

            i = j;
        }
        return final_state;
    }
};
